TypeCobol/Compiler/Diagnostics/CrossChecker.cs

```diff
@@ -464,6 +465,15 @@ public override bool Visit(DataDefinition dataDefinition)
 
         public override bool Visit(End end)
         {
+            // Check if PROGRAM END is orphan
+            if (_compilerOptions.CheckEndProgram.IsActive)
+            {
+                if (end.CodeElement.Type == CodeElementType.ProgramEnd && !(end.Parent is Program))
+                {
+                    DiagnosticUtils.AddError(end, "Unexpected orphan \"PROGRAM END\".", _compilerOptions.CheckEndProgram.GetMessageCode());
```

Comment:
Maybe add `Debug.Assert(end.Parent is SourceFile);` ?

Reply:
It's a good idea, it will make the potential maintenance easier.